torch/csrc/jit/runtime/graph_iterator.h

```diff
@@ -5,113 +5,196 @@ namespace jit {
-          // processing later.
-          block_stack_.emplace_back(current_.first, current_.second);
+        if (!else_block_empty) {
+          current_ = else_block->nodes().begin();
+        } else {
+          // Since it's empty we move to the next node.
+          current_ = owning_node_it;
+          ++current_;
+          if (current_ == owning_node_block->nodes().end()) {
+            move_up(*owning_node_it);
+          }
         }
+      }
+    } else if (
+        owning_node->kind() == prim::Loop ||
+        owning_node->kind() == prim::With) {
+      current_ = owning_node_it;
+      ++current_;
+      if (current_ == owning_node_block->nodes().end()) {
+        move_up(*owning_node_it);
+      }
+    } else {
+      throw std::runtime_error("Node should not have had any child blocks.");
+    }
+  }
 
-        if (!then_block_empty && else_block_empty) {
-          // Set current_ to {then_block, then_block.begin()} and push nothing
-          // to the stack since the else block is empty.
-          current_.first = then_block;
-          current_.second = then_block->nodes().begin();
-        } else if (then_block_empty && !else_block_empty) {
-          // Set current_ to {else_block, else_block.begin()} and push nothing
-          // to the stack since the current block is already on the stack.
-          current_.first = else_block;
-          current_.second = else_block->nodes().begin();
-        } else if (!then_block_empty && !else_block_empty) {
-          // Set current_ to {then_block, then_block.begin()} and push the
-          // else_block to the stack so that it will be processed after.
-          block_stack_.emplace_back(else_block, else_block->nodes().begin());
-          current_.first = then_block;
-          current_.second = then_block->nodes().begin();
-        }
-      } else if (node->kind() == prim::Loop || node->kind() == prim::With) {
-        auto* body_block = node->blocks().at(0);
+  // Moves to the next adjacent node or up in to the parent if that is not
+  // possible.
+  void move_next() {
+    auto block = current_->owningBlock();
+    Node* previous = *current_;
+
+    // Increment to the next node in the current block.
+    ++current_;
 
-        bool body_block_empty =
-            body_block->nodes().begin() == body_block->nodes().end();
+    // Check if we're at the end of the block. If so we need
+    // to move upwards (if it makes sense to).
+    if (current_ == block->nodes().end()) {
+      move_up(previous);
+    }
+  }
 
-        if (!body_block_empty) {
-          // If body_block is not empty, push the current block onto the stack
-          // to resume processing it later and set current_ to {body_block,
-          // body_block.begin()}.
-          block_stack_.emplace_back(current_.first, current_.second);
+  // Moves to the next node in the graph into children if it can.
+  void move_into() {
+    if (*current_ == nullptr) {
```

Comment:
This is more for safety/defensive programming and to prevent issues for future refactorings. Basically if we are in a position where we can't move further we shouldn't.

Reply:
could we make this an ASSERTION instead ? should it be an error if this fires ?

Reply:
Ack will do